### PR TITLE
Ensure mods are set before loading locale and database paths

### LINKED ISSUE #620: Locale and db not being read from mod path
There appears to be an issue with the mod loader, which blocks the locale from being read from the correct mod directory. base.po and base.mo are always loaded exclusively from mods/tuxemon and never from mods/<my_mod>

This appears to be caused by the locale code running before mods are set: I used a debug print to confirm that by the time locale.py calls fetch in prepare.py, the mods list only contains "tuxemon" and no other entries... only later calls to prepare.fetch are able to see both. The locale code should be executed after CONFIG.mods has been set.

### PR DESCRIPTION
**Pull request by MirceaKitsune**
Fixes #620

Due to the fact that locale and database paths were read instantly at startup, they did not have time to catch the mods being added to the mod list. As a result some resources could only be read from mods/tuxemon and never mods/<my_game> making the mod loaded unusable.

This required a bit of refractoring to locale and databases, which now receive the signal to load their data from pygame_init after the base settings have been set. As a bonus, this PR cleans up the database code and also removes several cases of databases being pointlessly loaded more than once.

I tested this as thoroughly as I could and am not seeing any issues: This PR loads fine for both stock Tuxemon as well as my custom game directory. Further testing is encouraged still.

### CODE REVIEW

**Comment by MirceaKitsune** [comment]
I was wondering if we could put this one on a fast track please. Unlike my other PR's which add new features or fix minor issues, this is a major bug fix. I'm working on my own game at the moment using the new mod system, and am presently only able to run it by using this branch.

I understand @bitcraft is busy and don't wish to rush him more than I have in the past. I was wondering if other developers are still available, and any one of you could review and merge this. @ShadowApex @pastaq @superman2k5 can you take a look at this and approve it please?

In terms of functionality I don't doubt it works: I've been using it for more than a day to constantly test changes to my mod all over the place. I never got any crashes, and missing file cases would almost always cause a crash at startup too. Only thing is that all subdirectories need to be present in your mod's db subdirectory (even if they're empty), also some hardcoded status techniques must be copied over... this is the case with the default mod too, mixing them is a feature for later.

**Comment by kerizane** [comment]
Looks fine, will do a quick test this evening.

**Comment by kerizane** [comment]
Error in cotton_town:

Python2
  File "/home/k/Tuxemon/tuxemon/core/event/actions/create_npc.py", line 76, in start
    sprite = databases.npc[slug].get('sprite_name')
TypeError: 'JSONDatabase' object has no attribute '__getitem__'

Python3
  File "/home/k/Tuxemon/tuxemon/core/event/actions/create_npc.py", line 76, in start
    sprite = databases.npc[slug].get('sprite_name')
TypeError: 'JSONDatabase' object is not subscriptable

File: tuxemon/core/event/actions/create_npc.py [hunk 1]
```diff
@@ -72,7 +73,7 @@ def start(self):
                 slug
             )
         else:
-            sprite = db.JSONDatabase('npc').database['npc'][slug].get('sprite_name')
```
Comment (kerizane) [inline tuxemon/core/event/actions/create_npc.py]: sprite = databases.npc.database['npc'][slug].get('sprite_name')

**Comment by kerizane** [comment]
`sprite = databases.npc.database['npc'][slug].get('sprite_name')`

This is pretty verbose, we should simplify it. 
Can be done in another ticket though

**Comment by kerizane** [comment]
Other than the create_npc crash, I'm happy.

I don't have merge rights though.

File: tuxemon/core/event/actions/create_npc.py (hunk 1 above)
Comment (MirceaKitsune) [reply tuxemon/core/event/actions/create_npc.py]: Really strange that I wasn't seeing it; I've been able to enter the maple house and talk to the professor's wife just fine. I tested your change and it works great so I committed it: Please double-check that this made the issue go away.

**Comment by MirceaKitsune** [comment]
Yeah, we should keep that simplification in mind for later. The rest of the database structure is the same, I only created a class to set and store all databases in one place; This was needed so pygame init could set them after mods are configured rather than each script (monster.py, techniques.py, etc) doing it instantly themselves... it's also cleaner and avoids some duplicate db reads.

File: tuxemon/core/event/actions/create_npc.py (hunk 1 above)
Comment (kerizane) [reply tuxemon/core/event/actions/create_npc.py]: Different code path, only affects some npcs

File: tuxemon/core/event/actions/create_npc.py [hunk 2]
```diff
@@ -72,7 +73,7 @@ def start(self):
...
+            sprite = databases.npc.database['npc'][slug].get('sprite_name')
```
Comment (bitcraft) [inline tuxemon/core/event/actions/create_npc.py]: this new format is extremely obtuse.  please find a way to make it more simple:
```python
db.npc['npc'][slug].get('sprite_name')
```
would be much better.

**Comment by bitcraft** [comment]
Overall, I'm not happy with this change.  I didn't expect a huge refactor of a core system without discussion first.  Please see my comment.  My primary objection is that the db system now requires remembering 3 names to access items.  Deeply nested structures like this should be avoided.  Please continue refactoring and make sure everything is tested.  You should be able to play until cotton town, have a monster, some items, and not crash.

**Comment by MirceaKitsune** [comment]
Latest commit removes the unnecessary level of indentation, using a single global database instead of an object spawning a new db instance for each field. I really should have seen this earlier.

File: tuxemon/core/event/actions/create_npc.py (hunk 2 above)
Comment (MirceaKitsune) [reply tuxemon/core/event/actions/create_npc.py]: Latest commit should fix this. It is now just `databases.database['npc']`, a single instance of JSONDatabase() is used as was most likely intended.

File: tuxemon/core/db.py [hunk 3]
```diff
@@ -86,11 +75,22 @@ def load(self, directory="all"):
 
         """
 
+        self.database = {
```
Comment (bitcraft) [inline tuxemon/core/db.py]: why is this moved?

File: tuxemon/core/db.py [hunk 4]
```diff
@@ -86,11 +75,22 @@ def load(self, directory="all"):
...
+            "item": {},
+            "monster": {},
+            "npc": {},
+            "technique": {},
+            "encounter": {},
+            "inventory": {},
+            "environment": {},
+        }
+
+        self.path = prepare.fetch("db")
```
Comment (bitcraft) [inline tuxemon/core/db.py]: instance attributes shouldnt be delcared outside of init.  python lets you do that, but its not a good practice.  please move "databases" back to init, and you can set path to None there.  leaving instance variables in init helps developers learn about the class and what it does.

File: tuxemon/core/db.py [hunk 5]
```diff
@@ -86,11 +75,22 @@ def load(self, directory="all"):
...
         if directory == "all":
             self.load_json("item")
             self.load_json("monster")
-            self.load_json("technique")
             self.load_json("npc")
+            self.load_json("technique")
```
Comment (bitcraft) [inline tuxemon/core/db.py]: why is this moved?

File: tuxemon/core/db.py [hunk 6]
```diff
@@ -192,3 +192,6 @@ def set_defaults(results, table):
                 )
 
     return results
+
+# Global database container
+databases = JSONDatabase()
```
Comment (bitcraft) [inline tuxemon/core/db.py]: `databases` is misleading.  this is a single object.  please change to `database`, or simply `db`

File: tuxemon/core/db.py (hunk 5 above)
Comment (MirceaKitsune) [reply tuxemon/core/db.py]: I noticed it wasn't in the same order as the database properties declared in init. "technique" comes after "npc" there, but below it was above it... I figured this would be more clean just in case.

File: tuxemon/core/db.py (hunk 3 above)
Comment (MirceaKitsune) [reply tuxemon/core/db.py]: I think I misunderstood one of our conversations on Discord. Placed it back in the last commit.

File: tuxemon/core/db.py (hunk 4 above)
Comment (MirceaKitsune) [reply tuxemon/core/db.py]: Fair enough, I shouldn't have missed that. I put it back in init as None... the load() function must set it for the mod issue to be solved, so only its configuration is done there now.

File: tuxemon/core/db.py (hunk 6 above)
Comment (MirceaKitsune) [reply tuxemon/core/db.py]: Renamed to db.

**Comment by MirceaKitsune** [comment]
Last commit tested briefly, no crashes or errors or missing content; I tried changing maps and battling and / or reading signs, for both vanilla tuxemon and my mod... both seem to be working perfectly now.

File: tuxemon/core/monster.py [hunk 7]
```diff
@@ -266,7 +262,7 @@ def load_from_db(self, slug):
         """
 
         # Look up the monster by name and set the attributes in this instance
-        results = monsters.lookup(slug)
+        results = db.lookup(slug)
```
Comment (kerizane) [inline tuxemon/core/monster.py]: results = db.lookup(slug, table="monster")

File: tuxemon/core/monster.py [hunk 8]
```diff
@@ -477,7 +473,7 @@ def set_flairs(self):
         if len(self.flairs) > 0 or self.slug == "":
             return
 
-        results = monsters.lookup(self.slug)
+        results = db.lookup(self.slug)
```
Comment (kerizane) [inline tuxemon/core/monster.py]: results = db.lookup(slug, table="monster")

File: tuxemon/core/monster.py (hunk 7 above)
Comment (MirceaKitsune) [reply tuxemon/core/monster.py]: Ouch... thanks for catching that.

File: tuxemon/core/monster.py (hunk 8 above)
Comment (MirceaKitsune) [reply tuxemon/core/monster.py]: Fixed.

**Comment by bitcraft** [comment]
Ok this looks like a good start.  there are still minor issues with it but they can be fixed in later PRs.  thanks again!
